### ISSUE TITLE
no-string-based-set-timeout produces false positives on inferred types

### ISSUE THREAD

**Issue by lddubeau** [microsoft/tslint-microsoft-contrib#355]
### Versions

tslint@4.5.1
tslint-microsoft-contrib@4.0.1
typescript@2.2.2

### Steps to reproduce

Using this `tslint.json`:

```json
{
    "rulesDirectory": "node_modules/tslint-microsoft-contrib",
    "rules": {
        "no-string-based-set-timeout": true
    }
}
```

this `tsconfig.json`:

```json
{
    "compilerOptions": {
        "target": "es6"
    }
}
```

and this `test.ts`:

```typescript
// This errors incorrectly.
new Promise((resolve) => setTimeout(resolve));

// This does not error.
const bar: (value?: {} | PromiseLike<{}>) => void = () => undefined;
setTimeout(bar);

// This errors incorrectly.
function foo(cb: (value?: {} | PromiseLike<{}>) => void): void {
    (function (x) {
        setTimeout(x);
    }(cb));
}

// This does not error.
function foo2(cb: (value?: {} | PromiseLike<{}>) => void): void {
    setTimeout(cb);
}
```

Using this command:

```
$ tslint --project tsconfig.json --type-check
```

### Actual Output

```
test.ts[2, 26]: Forbidden setTimeout string parameter: resolve
test.ts[11, 9]: Forbidden setTimeout string parameter: x
```

### Expected Output

No errors.

### Observations

It looks like the rule is unable to handle symbols with inferred types. The difference between `foo` and `foo2` is particularly revealing. If I query the type of `x` in `foo`, I get the same type as that of `cb` in `foo2`.

**Comment by lijunle** [comment]
It should been fixed in https://github.com/Microsoft/tslint-microsoft-contrib/pull/402 and target for version 5.0.2.

**Comment by JoshuaKGoldberg** [comment]
Adding to Junkern's comment, this is the error I get for the `private wait`:

```
 no-string-based-set-timeout    Forbidden setTimeout string parameter: resolve
```

**Comment by JoshuaKGoldberg** [comment]
Looks related to #442. Not sure if the same issue; needs a deeper investigation.

**Comment by schneems** [comment]
Is there a workaround for this? I'm super new to TS. I'm getting this failure `no-string-based-set-timeout  Forbidden setTimeout string parameter: resolve` with this code:

```ts
  async delay(ms: number) {
    return new Promise(resolve => setTimeout(resolve, ms))
  }
```

**Comment by JoshuaKGoldberg** [comment]
Welcome to TS, @schneems! 😊

As a workaround in the meantime, you can disable this rule on that one line with:

```ts
  async delay(ms: number) {
    // tslint:disable-next-line no-string-based-set-timeout
    return new Promise(resolve => setTimeout(resolve, ms))
  }
```

I just updated the issues labels to more accurately indicate that this is a bug we'd like to see fixed.

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is being deprecated**; therefore, it and tslint-microsoft-contrib are no longer accepting pull requests for major new changes or features. See https://github.com/palantir/tslint/issues/4534. 😱

If you'd like to see this change implemented, you have two choices:
* **Recommended:** Check if this is available in [ESLint](https://eslint.org) + [typescript-eslint](https://typescript-eslint.io) ✅
* _Not Recommended: Fork TSLint or tslint-microsoft-contrib locally_ 🤷‍♂️

👋 It was a pleasure open sourcing with you!

_If you believe this message was posted here in error, please comment so we can re-open the issue!_
